File: src/main/java/noteedit/service/FileService.java
package noteedit.service;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.core.conditions.update.UpdateWrapper;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import noteedit.entity.FileInfo;
import noteedit.mapper.FileMapper;
import noteedit.utils.FileUtils;
import noteedit.utils.JsonUtils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Slf4j
@Service
public class FileService extends ServiceImpl<FileMapper, FileInfo> {
    @Autowired
    private FileMapper fileMapper;

    /**
     * 将文件信息放进数据库
     * @param fileInfo 文件信息
     */
    public boolean linkFileToDatabase(FileInfo fileInfo) {
        log.info("linking files to database...");
        log.info("file info: " + fileInfo.toString());
        QueryWrapper<FileInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.eq("folder_id", fileInfo.getFolderId());
        List<FileInfo> files = fileMapper.selectList(queryWrapper);
        for(FileInfo file: files) {
            if(file.getFileName().equals(fileInfo.getFileName())) { // * 重名，返回false
                return false;
            }
        }
        fileMapper.insert(fileInfo);
        return true;
    }

    /**
     * 获取文件夹下的所有文件
     * @param folderId 文件夹id
     * @return 文件列表
     */
    public List<FileInfo> listAllFilesInFolder(Long folderId) {
        QueryWrapper<FileInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.eq("folder_id", folderId);
        return fileMapper.selectList(queryWrapper);
    }

    /**
     * 根据id获取文件信息
     * @param id 文件id
     * @return 文件信息
     */
    public FileInfo getFileById(Long id) {
        QueryWrapper<FileInfo> queryWrapper = new QueryWrapper<>();
        queryWrapper.eq("id", id);
        return fileMapper.selectOne(queryWrapper);
    }

    /**
     * 更新文件信息
     * @param fileInfo 新的文件信息
     * @return 是否更新成功
     */
    public boolean updateFileInfo(FileInfo fileInfo) {
        int affectedRows = fileMapper.updateById(fileInfo);
        return affectedRows == 1;
    }

    /**
     * 删除文件
     * @param id 文件id
     * @return 是否删除成功
     */
    public boolean deleteFileById(Long id) {
        FileInfo fileInfo = getFileById(id);
        int affectedRows = fileMapper.deleteById(id);
        if(affectedRows == 1) {
            FileUtils.deleteFile(fileInfo.getRealPath());
            return true;
        }
        return false;
    }
}

File: src/main/java/noteedit/service/FileService.kt
package noteedit.service

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl
import mu.KotlinLogging
import noteedit.entity.FileInfo
import noteedit.exception.NameAlreadyExistsException
import noteedit.mapper.FileMapper
import noteedit.utils.FileUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
open class FileService: ServiceImpl<FileMapper, FileInfo>() {
    private val logger = KotlinLogging.logger {}

    @Autowired
    private lateinit var fileMapper: FileMapper
    /**
     * 添加文件信息
     * @param fileInfo 文件信息
     * @throws NameAlreadyExistsException 文件夹下已存在同名文件
     * @return 是否添加成功
     */
    open fun add(fileInfo: FileInfo): Boolean {
        logger.info { "adding file $fileInfo ..." }
        require(
            fileMapper.selectList(
                QueryWrapper<FileInfo>()
                    .eq("folder_id", fileInfo.folderId)
                    .eq("name", fileInfo.name)
            ).isNullOrEmpty()
        ) { NameAlreadyExistsException("该文件夹下已存在同名文件") }
        return fileMapper.insert(fileInfo) == 1
    }

    /**
     * 获取指定文件夹下的所有文件信息
     * @param folderId 文件夹id
     * @return 文件列表
     */
    open fun listAllInFolder(folderId: Long): List<FileInfo> {
        logger.info { "listing all files in folder $folderId ..." }
        return fileMapper.selectList(
            QueryWrapper<FileInfo>()
                .eq("folder_id", folderId)
        ) ?: emptyList()
    }

    /**
     * 根据id获取文件信息
     * @param id 文件id
     * @return 文件信息
     */
    open fun getById(id: Long): FileInfo? {
        logger.info { "getting file $id ..." }
        return fileMapper.selectById(id)
    }

    /**
     * 更新文件信息
     * @param fileInfo 新的文件信息
     * @return 是否更新成功
     */
    open fun update(fileInfo: FileInfo): Boolean {
        logger.info { "updating file $fileInfo ..." }
        return fileMapper.updateById(fileInfo) == 1
    }

    /**
     * 删除文件信息
     * @param id 文件id
     * @return 是否删除成功
     */
    open fun delete(id: Long): Boolean {
        logger.info { "deleting file $id ..." }
        return (fileMapper.deleteById(id) == 1).also {
            getById(id)?.let {
                FileUtils.deleteFile(it.realPath)
            }
        }
    }
}